### ISSUE TITLE
VPC does not support Visual Studio 2015

### ISSUE THREAD

**Issue by myoKun345** [ValveSoftware/source-sdk-2013#325]
I have moved to using VS 2015, but unfortunately the "createallprojects.bat" file does not work for me because VPC does not have a mode for 2015. I was able to generate a 2013 solution by manually adding the "DefaultProjectExtension" registry values, but it does not work in VS 2015 because Visual Studio is seemingly terrible at converting solutions. Either that or the generated solution is missing needed elements of VS 2013 (beyond the extension registry values), which I do not have installed.

**Comment by yaakov-h** [comment]
Not surprising, VS2015 isn't released yet. :)

**Comment by myoKun345** [comment]
Haha :P

Effectively though, Microsoft is already pretty much pushing 2015 as the current version on the Visual Studio website and so on, and it is in the release candidate stage now.

**Comment by neico** [comment]
The real problem is that valve hardly ever looks in here anymore let alone update the repo, so my guess is we'll have to wait at least 2 years for such an update~

**Comment by geotavros** [comment]
Idk about the VS 2015 but all previous VS versions converted solutions 
very well. If you're having trouble generating the solutions you may 
want to try to install the VS 2013 Express and try again.

**Comment by dkreuter** [comment]
> I was able to generate a 2013 solution by manually adding the "DefaultProjectExtension" registry values.
Which registry? I though the way to make VPC generate vs2013 projects is by adding "/2013" to the line in createallprojects.bat

**Comment by aegookja** [comment]
>Which registry? I though the way to make VPC generate vs2013 projects is by adding "/2013" to the line in createallprojects.bat

should it be something like this?
devtools\bin\vpc.exe /hl2 /episodic +everything /2013 /mksln everything.sln

**Comment by dkreuter** [comment]
@aegookja exactly, I don't know if the order makes a difference.

**Comment by joeangry** [comment]
There seems to be a lot of issues with building the SDK with VS2015. Surely I'm not the only one that's experiencing these problems?

**Comment by P1x3lF3v3r** [comment]
You are not.

**Comment by Nerus87** [comment]
Same problem here.

**Comment by RaTcHeT302** [comment]
Did no one find a work around? I'd rather not install VS2013 again. Otherwise I guess I'll have to get this stuff setup with 2013 and then built it with 2015?

**Comment by psychonic** [comment]
Even if you get the project set up in VS2015, you'll still have to use the VS2013 (or lower) compiler tools. Due to ABI changes, the [precompiled static libs](https://github.com/ValveSoftware/source-sdk-2013/tree/master/mp/src/lib/public) are not compatible with VS2015.

**Comment by RaTcHeT302** [comment]
Meh, that sucks. But thanks anyway for telling me.

**Comment by geotavros** [comment]
I compile Alien Swarm mod with VS2013 but all its static libs are built with VS2005. It works fine for now. Do I risk to have undefined behavior? Should I compile with VS2005 to be on safe side?

**Comment by psychonic** [comment]
I believe you are alright, since most of the changes there affect ABI with the standard libraries. Valve uses that extremely sparingly, in favor of their own classes. VS 2015 shuffled around some built-in functions across libraries, which is what causes linker errors now, even if you tell it to bypass the library compiler version check with a define (which SDK 2013 already uses).

**Comment by bmk10** [comment]
no 2015, only 2013, and 2010. just cl.exe and its combatants

**Comment by Wazanator** [comment]
Bumping this because Microsoft is really making a push to get everyone on the latest version of Visual Studio by making it harder to find a download link.

A good portion of the comments on my tutorial on youtube and steam are in regards to Visual Studio or the problem they are having is because they used 2015

**Comment by joeangry** [comment]
I don't see this happening. The code does not compile with VS2015 anyway, and it'd require Valve to update their code.

**Comment by Foxtrek64** [comment]
I can sit down and sort through the code, see what needs to be fixed, but I'm assuming some issues lie within Source code only available with a VDN subscription. Still, if I can get this running, even if it's a bit hackish, it'll at least get us somewhere before the Source 2 SDK is released.

**Comment by Nerus87** [comment]
Anyway to fix this manually ?

**Comment by joeangry** [comment]
@Nerus87 there isn't. Valve needs to recompile some of their libraries with the 2015 compiler. Most of these libraries aren't in the SDK. @Nephyrin  any chance we could get the source for the libraries so we can take care of the SDK ourselves?

**Comment by Nerus87** [comment]
I think this engine R.I.P right now, Valve is not interested to continue improve or even fix bugs.

THX.

**Comment by StrikerMan780** [comment]
Microsoft has recently taken down all downloads of Visual Studio 2013 unless you're in the Dev Essentials program. This basically means the Source SDK is now impossible to develop with if you're a newcomer.

**Comment by psychonic** [comment]
The Dev Essentials program has no cost and takes less than 30 seconds to
sign up for if you already have a Microsoft account (which was already
required for the free VS 2013 editions).

Pretty impossible indeed.

Jordon Moss wrote:

**Comment by StrikerMan780** [comment]
I wasn't aware, when I tried to go to the page it was bugging out in my browser. I assumed it was a subscription-only thing like MSDN, or a program for registered businesses.

**Comment by malortie** [comment]
It appears that *Visual Studio Community 2013* is still be available.

You should be able to download it at [this address](https://www.microsoft.com/en-ca/download/confirmation.aspx?id=48129),
from official microsoft servers.

In addition, I uploaded the installer to my [Source SDK 2013 fork](https://github.com/malortie/source-sdk-2013), in case it completely
disappears from their servers.

[Visual Studio Community 2013 Update 5](https://github.com/malortie/source-sdk-2013/tree/master/vs2013/Update5)

I do hope this helps.

**Comment by kinnngg** [comment]
Will changing toolset work as told in [Source SDK 2013 Getting Started page](https://developer.valvesoftware.com/wiki/Source_SDK_2013)?
> Although Source SDK 2013 currently does NOT compile with Visual Studio 2015 (bug report) you can still use VS 2015 as IDE and compile the code with the VS 2013 toolset.

### [How to: Modify the Target Framework and Platform Toolset of Visual Studios](https://msdn.microsoft.com/en-us/library/ff770576.aspx)

**Comment by malortie** [comment]
@kinnngg 

Yes, but you will need to switch to the **v120** version of the toolset.

You require to have the **v120** toolset installed. If not, you will need to download it from their servers. Once this is done, open up **games.sln** or **everything.sln** with the IDE of your choice.

When you open up for the first time, a project that was created in a previous version of VS, you will be prompted to update it's toolset. In this case, you should choose Not to update it.

In case you updated projects, then open up the properties (right click in solution explorer), and in the main configuration layout, find **Platform Toolset**, it should be *v140* in *2015* and *v141* if you are using the *2017 RC* version. In the list there should be **Visual Studio 2013 (v120)**. Select it.

You should then be able to compile it.

**Comment by kinnngg** [comment]
I have chosen "Not to update" but it's showing errors when I try to compile in VS2015.
How can I change toolset coz there is no option to change Platform Toolset in Configuration Menu :(

Thanks

**Comment by malortie** [comment]
@kinnngg 

Do you have Visual Studio 2013 installed? If not, the toolset cannot be found, which is the reasons you are experiencing issues.

Unfortunately, it was removed from their servers. The only way to have it is to obtain it via Dev essentials, as stated by users above.

**Comment by kinnngg** [comment]
@malortie 
I thought Windows 8.1 & Windows Phone 8.1 SDK includes v140_xp 😕 
Thanks. I gonna download it.

I already have VS2015, 2010, 2012 installed 😆 Now its time for 2013 OMG 😉

**Comment by joeangry** [comment]
If you have 2012 you should be able to compile using that toolset.

**Comment by kinnngg** [comment]
@jeoangry
Tried bro. But didn't work.

**Comment by kukkerman** [comment]
Hi!

I've managed to create _mp/src/everything.sln_ for VS2015 with some registry magic. After some refactoring every project compiles fine except:

- _Phonemeextractor_ (I don't have speech SDK installed)
- _QC_eyes_ (I don't have MFC installed)

I had to modify many of the source files, because of the /W4 (warning level 4) and /WX (treat warnings as errors) compiler switches and because VS2015 being more syntax sensitive (especially for variables hiding another from an outer scope).
The most obscure linking error required dissecting _mp/src/lib/public/particles.lib_, renaming the exported symbol __hypot_ to __hypot_particles_ and reassembling it again. It was necessary because __hypot_ is already defined in the standard C runtime library and the linker cannot choose automatically between them.

If somebody is still interested in working with source engine under VS2015 let me know, and I'll publish my modifications in my github repo.

Cheers!

**Comment by joeangry** [comment]
I managed to it working using 2017 because 2017 can install the 2013 toolset. So that'd be another way of doing it.

**Comment by Nerus87** [comment]
Nice to know, thx.

**Comment by TomyLobo** [comment]
@joeangry I was dazed and confused when that [link in the wiki](https://developer.valvesoftware.com/wiki/Source_SDK_2013#Step_One:_Installing_Visual_Studio) actually gave me a working vs community 2013 installer. Got everything working with that.
So yay, finally a recent IDE to develop Double Action in :)

**Comment by BlackPhrase** [comment]
@kukkerman 
Nice job!
Yep, I'm interested in making this thing buildable by MSVC2015/2017+
Did you manage to make it buildable by MSVC2015?
Are compiled game dlls playable?
What have you done with linking errors?

**Comment by StrikerMan780** [comment]
Any way to install the VS2013 toolset in VS2019?

**Comment by Creepysin** [comment]
@StrikerMan780 You just install VS2013 like normal and it should just work in VS2019 after that.

**Comment by GamerDude27** [comment]
@StrikerMan780 Yeah, as long as you are using the `v120_xp` toolset you can use VS 2019. That's what I'm doing.

**Comment by TylerJaacks** [comment]
@GamerDude27 So you were able to fully compile and build with VS 2019 with v120_xp. How exactly do you install v120_xp for VS 2019 I though they removed support for this?

**Comment by malortie** [comment]
@TylerJaacks 

You must install Visual Studio 2013, which will install the tools, headers, libraries to build C++ applications using the v120_xp platform toolset.

On the [Source SDK 2013 VDC Page](https://developer.valvesoftware.com/wiki/Source_SDK_2013#Step_One:_Installing_Visual_Studio), there is a download link for VS2013 that appears to be working, so you should be able to install VS2013. If you already have Visual Studio 2019, I recommend that you uninstall it first, then install Visual Studio 2013, and then install Visual Studio 2019 again. After you installed VS2013, run one of the batch files:

- createallprojects.bat -> will create a solution file named everything.sln
- creategameprojects.bat -> will create a solution file named games.sln

Open games.sln using Visual Studio 2019. When prompted to choose whether or not to migrate projects to newer platform toolset, select 'No' so that the projects remain on the Visual Studio 2013 (v120_xp) toolset.

You should be able to build the projects normally while using Visual Studio 2019 IDE.

**Comment by TylerJaacks** [comment]
Can confirm that it does work I all need was make sure 2013 C++ was installed and I could do all my development in 2019 or even VS Code.

**Comment by AbrahamBrookes** [comment]
That download link on the valve developers page downloads an installer, and that installer can no longer locate the files it needs to download! There is an ISO hosted on my.visualstudio.com (requires a microsoft account) https://my.visualstudio.com/Downloads?q=visual%20studio%202013%20with%20update%205

**Comment by halamix2** [comment]
@kisak-valve : sorry for bothering you, but I feel like this issue could be closed, since Source SDK 2013 now requires Visual Studio 2022

**Comment by Jesse-Culver** [comment]
> [@kisak-valve](https://github.com/kisak-valve) : sorry for bothering you, but I feel like this issue could be closed, since Source SDK 2013 now requires Visual Studio 2022

The single player branch has not been updated to include support for 2022 from what I can tell looking at the commit history.

**Comment by halamix2** [comment]
> The single player branch has not been updated to include support for 2022 from what I can tell looking at the commit history.

You're right, I've overlooked the `singleplayer` branch completely.


So the status is:
* `master` requires VS 2022
* `singleplayer` requires VS 2013; there was an attempt made to port it to VS 2022[^1], but it seems abandoned.

[^1]: #975

**Comment by tschumann** [comment]
Out of curiosity, does using the updated vpc binary work?

**Comment by halamix2** [comment]
Note: Visual Studio 2026 was released
When running `createallprojects.bat` the .sln project created for VS 2022 works well enough with VS 2026 to compile everything.

Running TF in Release mode works

When running TF in Debug it starts to the main menu and fails an assertion at `src\tier1\utlbuffer.cpp:222`
 and (unsurprisingly) segfaults at `src\game\shared\econ\econ_item.cpp:483` when assert is ignored

<img width="404" height="289" alt="Image" src="https://github.com/user-attachments/assets/32796c47-464b-44c2-857b-db6cda7b90b4" />
